jscomp/ml/ctype.ml

```diff
@@ -3951,6 +3951,30 @@ let rec subtype_rec env trace t1 t2 cstrs =
         end
     | (Tconstr(p1, _, _), _) when generic_private_abbrev env p1 ->
         subtype_rec env trace (expand_abbrev_opt env t1) t2 cstrs
+    | (Tconstr(_, [], _), Tconstr(p, [], _)) when Path.same p Predef.path_string -> (* type coercion for variants represented by strings *)
```

Comment:
This is not strictly true: it's between type constructors and strings.
So this will catch all such cases, override the default handling, and prevent other coercions to string in the next lines.
I think the entire guard should mode to Variant_coercion, so here only the highest level logic is left.

Reply:
Check the latest commit and see what you think.